metadeploy/api/views.py

```diff
@@ -176,52 +191,160 @@ def filter_get_one(self, qs):
+
+        plan = self.get_object()
+        valid_plan_type = (
+            plan.supported_orgs == SUPPORTED_ORG_TYPES.Scratch
+            or plan.supported_orgs == SUPPORTED_ORG_TYPES.Both
+        )
+        if not valid_plan_type:
+            return Response(
+                {"detail": "This plan does not support creating a scratch org."},
+                status=status.HTTP_409_CONFLICT,
+            )
+
+        new_data = request.data.copy()
+        new_data["plan"] = str(plan.pk)
+        serializer = ScratchOrgJobSerializer(data=new_data)
+        if not serializer.is_valid():
+            return Response(serializer.errors, status=status.HTTP_400_BAD_REQUEST)
+
+        # Check queue status
+        # If overfull, return 503
+        queue = django_rq.get_queue("default")
+        if len(queue) > settings.MAX_QUEUE_LENGTH:
+            return Response(
+                {"detail": "Queue is overfull. Try again later."},
+                status=status.HTTP_503_SERVICE_UNAVAILABLE,
+            )
+
+        uuid = str(uuid4())
+        scratch_org = serializer.save(uuid=uuid)
+        request.session["scratch_org_id"] = uuid
+
+        serializer = ScratchOrgJobSerializer(instance=scratch_org)
+        return Response(
+            serializer.data,
+            status=status.HTTP_202_ACCEPTED,
+        )
+
+    @action(detail=True, methods=["post", "get"], permission_classes=(AllowAny,))
+    def scratch_org(self, request, pk=None):
+        if request.method == "GET":
+            return self.scratch_org_get(request)
+        if request.method == "POST":
+            return self.scratch_org_post(request)
+
 
 class OrgViewSet(viewsets.ViewSet):
+    permission_classes = (AllowAny,)
+
     def list(self, request):
         """
         This will return data on the user's current org. It is not a
         list endpoint, but does not take a pk, so we have to implement
         it this way.
         """
-        current_job = Job.objects.filter(
-            org_id=request.user.org_id, status=Job.Status.started
-        ).first()
-        current_preflight = PreflightResult.objects.filter(
-            org_id=request.user.org_id, status=PreflightResult.Status.started
-        ).first()
-        serializer = OrgSerializer(
-            {"current_job": current_job, "current_preflight": current_preflight}
-        )
-        return Response(serializer.data)
+        # TODO: make this return { anonymous: {}, nonymous: {} }
+        # TODO: make the ScratchOrgJob:uuid channel get events from
+        # related Jobs and PreflightResults
+        response = {}
+
+        if "scratch_org_id" in request.session:
+            uuid = request.session["scratch_org_id"]
+            current_job = Job.objects.filter(
+                uuid=uuid, status=Job.Status.started
+            ).first()
+            current_preflight = PreflightResult.objects.filter(
+                uuid=uuid, status=PreflightResult.Status.started
+            ).first()
```

Comment:
@wlonk For these to work, don't we need to set the `uuid` on Jobs and Preflights when they're created from a ScratchOrg?

I'm not entirely sure why Jobs and Preflights need uuids to begin with. Couldn't we set `org_id` on ScratchOrg, use `uuid` only to look up the relevant ScratchOrg, then use `scratch_org.org_id` to lookup Jobs/Preflights as needed?

Reply:
We were setting it on jobs, but that got commented out and we haven't re-implemented triggering jobs yet. Added it to preflights.

Sure, we could do that. I have no strong opinions.

Reply:
I think if we can isolate the differences between ScratchOrgs and PersistentOrgs to the orgs themselves, such that other models (Jobs/Preflights) interact with them similarly, that's probably cleaner going forward?